accepted/single-file/extract.md

```diff
@@ -4,112 +4,115 @@ When running the single-file, the host will need to extract some bundled files r
-</PropertyGroup>
-```
+  Each single-file bundle contains a path-compatible cryptographically strong identifier embedded within it. This identifier is used as part of the extraction path in order to ensure that files from multiple versions of an app do not interfere with each other.
 
-An app may choose to have all of its dependencies extracted to disk at runtime (instead of loading certain files directly from the bundle) by setting the following property.
+#### Cleanup
 
-```xml
-<PropertyGroup>
-    <SingleFileExtractAll>true</SingleFileExtractAll>
-</PropertyGroup>
-```
+The cleanup of extracted files in the install-location will be manual in this version. 
 
-## Proposed Solution
+We can consider providing helper scripts to cleanup the extracted files for an executable. However, future versions are expected to spill fewer artifacts to disk, making the cleanup commands a lower priority feature.
+
+### Extraction Mechanism
+
+During startup, the Apphost:
+
+- Checks if pre-extracted files are available for the app, and if so, reuses them.
+  - In this version, the extracted files are assumed to be correct by construction. That is, if  `<base>/<app>/<bundle-id>`  exists, the files there are reused.
+  - If the contents of the extraction directory are corrupted post-extraction, they will need to be explicitly cleaned.
+  - This situation is no different from having a managed app with one of the dependent DLLs corrupted.
+- Extracts appropriate files to a temporary location `<base>/<app>/<proc-id>` . This ensures that concurrent launches of the app do not interfere with each other.
+- After successful extraction of all necessary files to `<base>/<app>/<proc-id>`, the app host commits the extraction to the actual location `<base>/<app>/<bundle-id>` .
+  - If `<base>/<app>/<bundle-id>` already exists by now, another process may have completed the extraction, so the host removes the directory `<base>/<app>/<proc-id>`, and continues execution with files in `<base>/<app>/<bundle-id>` .
+  - If not,  `<base>/<app>/<proc-id>` is renamed to `<base>/<app>/<bundle-id>` 
+- The `<base>/<app>/<proc-id>`  is now considered app-root, and execution continues.
+
+### Extraction Configurations
+
+The following environment variables influence the extraction mechanism:
 
-For .Net Core 3.0, we propose that we start with the simple implementation that extracts necessary files to temporary locations on every run at startup. Based on customer feedback, we can implement other options discussed above.
\ No newline at end of file
+* `DOTNET_BUNDLE_EXTRACT_BASE_DIR`  The base directory within which the files embedded in a single-file app are extracted, as explained in sections above. This directory can be set up an admin-only writable location if necessary.
+* `DOTNET_BUNDLE_EXTRACT_ALL`:  Apps may choose to have all of their dependencies extracted to disk at runtime (instead of loading certain files directly from the bundle) by setting this variable.
```

Comment:
What is the scenario when/how the app will use this environment variable? If the app wants to choose, shouldn't it be a config setting instead?

Reply:
I think it is useful to have settings in both forms:
* A option in `runtimeconfig.json`, since this is the best way to configure shipping apps.
* An environment variable, to assist in debugging scenarios, or if the customer wants to explicitly control where all extracted files land.

In the current preview, all embedded files are spilled to disk upfront, even before configuration files are processed. Therefore, we only have the environment variables. In the upcoming version, configuration files will be processed directly from the bundle. I'll add runtimeconfig options once this is done.

Reply:
@jkotas does this address your concern? Thanks.

Reply:
The design docs should capture what we plan to build, not the point-in-time state. I think doc should say that the environment variable is meant just for debugging. The design doc suggests that it is something we would recommend to deploy apps with. I do not think we want to recommending the apps with this.

BTW: What happens when I run the app without the env variable (only the non-bundleable files get expanded) and then run it again with env variable? Is the second run going to crash because of the app will expect all files to be on disk, but they won't be there?

Reply:
OK Thanks @jkotas, I'll update the doc with config settings.

DOTNET_BUNDLE_EXTRACT_ALL means that all the files are always be extracted.
I think this flag is only useful for debugging.

Reply:
@jkotas: I've addressed the feedback in this change: 2b7f4a6d841de30581e8dfdbb8aca4e39d968b17 Thanks.